### ISSUE TITLE
Consider adding column containing server address into system.query_log table

### ISSUE THREAD

**Issue by PhantomPhreak** [ClickHouse/ClickHouse#3994]
We want to keep a query log from all servers in a cluster in one table.
To do this, we're using replicated tables and distributed table on the top, and following server config
```
    <query_log>
        <database>_database_</database>
        <table>_distributed_query_log_table_</table>
    </query_log>
```
It's pretty simple and makes it possible to analyze requests from the entire cluster, with a tools like [this](https://grafana.com/dashboards/2515), or [this](https://grafana.com/plugins/vertamedia-clickhouse-datasource)

Problem: since current `system.query_log` doesn't contain a column with a server address, we can't figure out on which server the particular request was executed.

As a workaround, we can make a materialized view from `system.query_log` local tables, inserting data to distributed table, but it will be more useful to add one more column with a server address.

Thanks!

**Comment by filimonov** [comment]
You can just use [hostName](https://clickhouse.yandex/docs/en/query_language/functions/other_functions/#hostname) function.

BTW: If you will create Distributed table above the querylog in the same cluster as your production data - you will get logs only from **one** replica in each shard. Alternatives are:
1) configure extra cluster with all replicas as shards
2) use [remote](https://clickhouse.yandex/docs/en/query_language/table_functions/remote/) table function

**Comment by PhantomPhreak** [comment]
> You can just use [hostName](https://clickhouse.yandex/docs/en/query_language/functions/other_functions/#hostname) function.
I haven't found a way how to use it in this case, can you please make an example?

> If you will create Distributed table above the querylog in the same cluster as your production data - you will get logs only from **one** replica in each shard

Yes, it's a logic of tables with `Distributed` engine. Since we're using 2 replicas in each shard, we're using replicated tables for keeping logs, and table with `Distributed` engine on the top.

> 1. configure extra cluster with all replicas as shards
Yes, it works, but defining additional cluster just for this feature looks excessive.

> 2\. use [remote](https://clickhouse.yandex/docs/en/query_language/table_functions/remote/) table function
It's a good idea, i will try it, thanks!

**Comment by filimonov** [comment]
> I haven't found a way how to use it in this case, can you please make an example?

```
SELECT *, hostName() as host_name FROM distributed;
```

**Comment by PhantomPhreak** [comment]
> > I haven't found a way how to use it in this case, can you please make an example?
> 
> ```
> SELECT *, hostName() as host_name FROM distributed;
> ```
> > . Since we're using 2 replicas in each shard, we're using replicated tables for keeping logs, and table with Distributed engine on the top.
> 
> But you probably want to collect querylog from all the 4 servers, don't you?

Exactly. And i'd like to see a server, executed a query, logged in `query_log` table, not a host handled `SELECT * FROM query_log_distributed` request.
So, this won't solve my problem, unfortunately

**Comment by filimonov** [comment]
> And i'd like to see a server, executed a query, logged in query_log table, not a host handled SELECT * FROM query_log_distributed request.
> So, this won't solve my problem, unfortunately

Did you try? :) [Documentation](https://clickhouse.yandex/docs/en/query_language/functions/other_functions/#hostname) explicitly mention that case:

> ### hostName()¶
> Returns a string with the name of the host that this function was performed on. **For distributed processing, this is the name of the remote server host,** if the function is performed on a remote server.

**Comment by PhantomPhreak** [comment]
Yes, i tried. It works, so i solved my problem in following way:
1. Created replicated tables with the same schema as `system.query_log` table, but with additional column `hostname` 
2. Created materialized view consuming from local `system.query_log` tables, adding `hostname` field and result of calling `hostname()` function, and storing result to replicated tables created in step #1

But anyway, having "hostname" field would be more convinient and will allow to omit creating materialized view

**Comment by azat** [comment]
>Exactly. And i'd like to see a server, executed a query, logged in query_log table, not a host handled SELECT * FROM query_log_distributed request.
>So, this won't solve my problem, unfortunately

You can use [`_shard_num`](https://clickhouse.tech/docs/en/engines/table-engines/special/distributed/#virtual-columns) for this:

```sql
select c.host_name, * from query_log_distributed inner join system.clusters c on c.shard_num = _shard_num and c.cluster = 'cluster_of_query_log_distributed`
```

But it has some issues in subqueries (#12601 ), and the workaround for now is to use subquery.

Refs: #7624 (19.17.4.11+)

**Comment by UnamedRus** [comment]
Done in https://github.com/ClickHouse/ClickHouse/pull/55894
